api/src/main/java/org/xnio/ByteBufferSlicePool.java

```diff
@@ -155,15 +155,11 @@ private Slice allocateSlices(final int buffersPerRegion, final int bufferSize) {
         // only true if using direct allocation
         if (directBuffers != null) {
             ByteBuffer region = FREE_DIRECT_BUFFERS.poll();
-            try {
-                if (region != null) {
-                    return sliceReusedBuffer(region, buffersPerRegion, bufferSize);
-                }
-                region = allocator.allocate(buffersPerRegion * bufferSize);
-                return sliceAllocatedBuffer(region, buffersPerRegion, bufferSize);
-            } finally {
```

Comment:
If this code is meant to only add the region on failure (my memory isn't clear but that's what the JIRA says), then shouldn't the fix be to change the `finally` to a `catch (Throwable t)` + re-`throw`, and keep the re-add logic so the region is available to the next allocation?

Reply:
@dmlloyd you're right about it, I'll update the PR

Reply:
@dmlloyd as a matter of fact, the reason I was not catching it is that it is not safe to readd the region for the next allocation.
The code invoked (sliceAllocatedBuffer and sliceReusedBuffer) might had added buffers to the queue before the exception was thrown. If we mark the region for reuse, we risk providing the same allocated memory to more than one allocate request, because we are marking the region for slicing again in the future

Reply:
And, now I recall, I added the misleading description to the Jira before realizing that it was not safe marking the region for reuse in the case of an exception. I hope this clarifies why it is different from the PR I provided. I'll update the description.

Reply:
@dmlloyd because I need to do the release, I'll merge this as is (the bug can cause serious failures of concurrent access to the same buffer) and then we can discuss this tomorrow. If the fix needs improvement, I'll open a new Jira and submit a PR in the sequence

Reply:
@dmlloyd it turns out there is nothing wrong with the original code, this commit introduced a new bug: directBuffers is just used to keep track of all direct memory allocated so it can be added to FREE_DIRECT_BUFFERS
I'll have to revert this commit, sorry for the big confusion